### PR TITLE
feat: add ability to log benchq params to mlflow

### PR DESCRIPTION
**Pull request by 1ethanhansen**
## Description

Include description of feature this PR introduces or a bug that it fixes. Include the following information:

- Context: Users often want to track and share information about and resulting from benchq resource estimates. [mlflow](https://mlflow.org/) is a good tool for that.
- In order to make it easier for users to log the values they care about to mlflow, new helper functions are added for logging input objects (`AlgorithmImplementation`, `BasicArchitectureModel`, and `DecoderModel`) to mlflow, as well as logging info from the output `ResourceInfo` object)
- mlflow is added as a dependency. This is needed in order to be able to call mlflow logging functions

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation. (new example added showing use of the functions)

### CODE REVIEW

File: examples/ex_6_mlflow.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    BASIC_SC_ARCHITECTURE_MODEL,
+    AlgorithmImplementation,
+    ErrorBudget,
+    get_program_from_circuit,
+)
+from benchq.resource_estimation.graph import (
+    GraphResourceEstimator,
+    create_big_graph_from_subcircuits,
+    run_custom_resource_estimation_pipeline,
+    simplify_rotations,
```
Comment (max-radin) [inline examples/ex_6_mlflow.py:21]: `simplify_rotations` was recently renamed `transpile_to_native_gates`

File: src/benchq/mlflow/data_logging.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    ResourceInfo,
+)
+
+
+def log_input_objects_to_mlflow(
+    algorithm_description: AlgorithmImplementation,
+    algorithm_name: str,
+    hardware_model: BasicArchitectureModel,
+    decoder_model: Optional[DecoderModel] = None,
+):
```
Comment (max-radin) [inline src/benchq/mlflow/data_logging.py:24]: Add `-> None` to indicate return value type

File: src/benchq/mlflow/data_logging.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    mlflow.log_params(_flatten_dict(asdict(hardware_model)))
+
+    # because decoder model is optional, check to make sure it exists
+    if decoder_model is not None:
+        mlflow.log_params(_flatten_dict(asdict(decoder_model)))
+    else:
+        mlflow.log_param("decoder_model", "None")
+
+
+def log_resource_info_to_mlflow(resource_info: ResourceInfo):
```
Comment (max-radin) [inline src/benchq/mlflow/data_logging.py:47]: Add `-> None` to indicate return value type

**Review by max-radin — CHANGES_REQUESTED** [review]
Looks good! I was able to run the example and see the experiments in mlflow.

Left a few minor comments. Also since there is already is an example 6, we should rename this to example 7.

**Comment by 1ethanhansen** [comment]
> Looks good! I was able to run the example and see the experiments in mlflow.
> 
> Left a few minor comments. Also since there is already is an example 6, we should rename this to example 7.

Oh that example must have gotten added recently! Changed to 7 in 0cb9d646fb82f9c4f5add0f7ad8411a7f40b9a3c

File: src/benchq/mlflow/data_logging.py (hunk 2 above)
Comment (1ethanhansen) [reply src/benchq/mlflow/data_logging.py]: fixed in ba87b5c7cf219232747b7e560dae8ec237e610c5

File: src/benchq/mlflow/data_logging.py (hunk 3 above)
Comment (1ethanhansen) [reply src/benchq/mlflow/data_logging.py]: fixed in ba87b5c7cf219232747b7e560dae8ec237e610c5

**Review by max-radin — APPROVED** [review]

**Pull request merged**